### ISSUE TITLE
nextSibling issue

### ISSUE THREAD

**Issue by dy** [luwes/js-diff-benchmark#16]
```js
  let parent = new Dommy()
  const childNodes = [];
  for (let i = 0; i < 100; i++) childNodes.push(new Nody(parent, i));

  console.log(childNodes[0].nextSibling)  // here it's end node
  parent.appendChild(childNodes[1])
  console.log(childNodes[0].nextSibling) // here it's childNodes[1]
```

**Comment by luwes** [comment]
not ideal, is it resulting in wrong behavior in the benchmark?

**Comment by dy** [comment]
It breaks some algos that depend on checking `nextSibling`, like swapping items etc.

**Comment by dy** [comment]
In particular, than bug doesn't allow [this 235b](https://github.com/spectjs/spect/blob/h-reducers/h.js#L288-L318) implementation to be qualified:
```js
export function merge (parent, a, b, before) {
  const bs = new Set(b), as = new Set(a)
  let i, ai, bi, off

  // walk by b from tail
  // a: 1 2 3 4 5, b: 1 2 3 → off: +2
  for (i = b.length, off = a.length - b.length; i--;) {
    ai = a[i + off], bi = b[i]
    // console.log(ai, bi, [...parent.childNodes].map(child => child.data))

    if (ai === bi) {}

    // replace
    else if (ai && !as.has(bi) && !bs.has(ai)) (parent.replaceChild(bi, ai))

    // remove
    else if (ai && !bs.has(ai)) (parent.removeChild(ai), off--, i++)

    else if (bi.nextSibling != before || !bi.nextSibling) {
      // swap
      if (as.has(bi)) (parent.insertBefore(ai, bi), off--)

      // insert
      parent.insertBefore(bi, before), off++
    }

    before = bi
  }

  return b
}
```
In real dom it works brilliant.
I couldn't manage to figure out why there's `_childNodes` with extra element and fix that.

**Comment by dy** [comment]
Interestingly, list-difference is broken in real DOM due to that reason.

**Comment by dy** [comment]
Ok, just in case - reproduced benchmark in DOM fully: https://github.com/spectjs/spect/blob/h-reducers/test/diff.js
Libs seems to be broken on real DOM, unfortunately.

**Comment by luwes** [comment]
great find, which ones are broken?

**Comment by dy** [comment]
I've tested `list-difference`, `snabbdom`, `udomdiff` - all algos seem to fail on real DOM (I've just reproduced benchmark tests). It's due to that `childNodes.lastElement` thing I suppose.

**Comment by dy** [comment]
Just in case - there are multiple various DOM implementations for nodejs:
* [html-document](https://ghub.io/html-document)
* [basichtml](https://ghub.io/basichtml)
* [dom-lite](https://ghub.io/dom-lite)
* [jsdom](https://ghub.io/jsdom)
* [happy-dom](https://ghub.io/happy-dom)
* [undom](https://ghub.io/undom)
* [simple-dom](https://ghub.io/simple-dom)
* [micro-dom](https://ghub.io/micro-dom)
* [min-document](https://github.com/Raynos/min-document)
* [nodom](https://github.com/redom/nodom)

**Comment by luwes** [comment]
hmm that would be strange, that all would be failing.

yea good idea, I haven't thought of using a more robust DOM library.
I'll look into some. I guess we would want to keep the operation count but that could be patched on maybe.

**Comment by luwes** [comment]
ah I see you removed the `before` node completely. yes I think some algorithms really rely on that to be there.

**Comment by luwes** [comment]
also wanted to mention sometimes spect fails the assertion of the shuffled test.
not a lot but once every 3 or 4 times.

<img width="453" alt="Screen Shot 2020-04-16 at 7 40 44 PM" src="https://user-images.githubusercontent.com/360826/79516764-4f876600-801a-11ea-82fd-77ee310d9520.png">

**Comment by dy** [comment]
Thanks for letting know. I caught similar error rewriting the thing today, basically `nextSibling` array was error-prone. The 232b version doesn't have that issue:
![image](https://user-images.githubusercontent.com/300067/79519349-9cbb0600-8021-11ea-9fcf-e941578a4324.png)

**Comment by WebReflection** [comment]
> I couldn't manage to figure out why there's `_childNodes` with extra element and fix that.

As the benchmark was born after a file which only purpose was to code-cover synthetically _udomdiff_, benchmark that simply reproduces what _js-framework-benchmark_ showcases, the last element is needed to *pin* the operation in a specific point.

Here how it works:

```html
<body>
  <!--pin-->
</body>
```

That *pin* comment node, makes it possible to confine all operations *before* it.

```js
var newNodes = udomdiff(
  document.body,
  [],
  [document.createElement('a'), document.createElement('b')],
  pin
);
```

After this operation, the layout would look like:
```html
<body>
  <a></a><b></b>
  <!--pin-->
</body>
```

The *pin* basically confines operations because other wise, if a node that should be inserted but it's not live yet, uses `nextSibling`, and the value is `null`, the `insertBefore` operation would append that node at the end of the parent container, and the end can be after any other node.

_udomdiff_ is a standalone library that covers all real-world uses cases I could try, so if there's a scenario where it fails in the real DOM, but again, it's heavily tested against that, an issue filed in _udomdiff_ repository would be much better than discussing in here the Dommy facade, as that's, once again, there only to synthetically quickly test and benchmark the library, it was never meant to be used across multiple libraries.

You can try [basichtml](https://github.com/WebReflection/basicHTML#readme) or [jsdom](https://github.com/jsdom/jsdom#readme) instead, if you want a more reliable DOM env, or you can rewrite Dommy as such:

```js
class Dommy extends Array {
  constructor() {
    super().operations = [];
  }
  get childNodes() {
    return this.slice(0);
  }
  get textContent() {
    return this.map(node => node.value).join('');
  }
  insertBefore(newNode, liveNode) {
    if (!liveNode)
      this.push(newNode);
    else {
      let i = this.indexOf(newNode);
      if (-1 < i)
        this.splice(i, 1);
      i = this.indexOf(liveNode);
      if (i < 0)
        throw new Error('invalid insertBefore');
      this.splice(i, 0, newNode);
    }
  }
  replaceChild(newNode, oldNode) {
    let i = this.indexOf(newNode);
    if (-1 < i)
        this.splice(i, 1);
    i = this.indexOf(oldNode);
    if (i < 0)
        throw new Error('invalid replaceChild');
    this[i] = newNode;
  }
  removeChild(node) {
    let i = this.indexOf(node);
    if (i < 0)
      throw new Error('invalid removeChild');
    this.splice(i, 1);
  }
  // count() => this.length
  // reset() => this.splice(0);
}
```

At that point you can either do an initial push of a comment node and pass it along, verifying that it never gets removed and it's always the last element, or you can just drop the `before` part, as this particular benchmark doesn't need it.

I would prefer to keep the *pin* concept, as diffing without that ability, in the *real-world* is kinda useless, 'cause handling all nodes per each container, is not really how libraries applied to the DOM work.

**Comment by dy** [comment]
> The pin basically confines operations because otherwise, if a node that should be inserted but it's not live yet, uses nextSibling, and the value is null, the insertBefore operation would append that node at the end of the parent container, and the end can be after any other node.

I see. In other words, to differentiate detached element from the last element in container in eg. `el.nextSibling === end` case. Btw that's not mandatory for all algos - some do just fine considering `end === null` as the "end of container", same as `el.insertBefore(child, null)`. That's mixing of concerns - figuring out `node.parentNode === currentParent` and `node.nextSibling === null`, fixing it with lies may give some relief, but in general relying on presence of `endPin` is wrong - if needed, algorightm may create _pin_ itself.

I tried rewriting Dommy for both problems - refs for `nextSibling/previousSibling` instead of `indexOf` and getting rid of `_childNodes`, with little success.

> diffing without that ability, in the real-world is kinda useless

I'd rather disagree: diffing all child nodes of a container is what react/preact/morphdom/nanomorph/etc do: `render(children, container)`, and the last element is implied `null` here, so algos should be just fine without last pin. 

I see you fixed that in the #19, great!

**Comment by WebReflection** [comment]
> Btw that's not mandatory for all algos

I've removed the `before` from the equation, it's optionally enabled on top of the file now.

> `render(children, container)`

same goes for _uhtml_, `render(where, what)`, but when it maps a template literal, it needs to know where to work, and never append anything after, example:

```js
render(document.body, html`
  <div>
    ${items.map(item => html`<p>${item.value}</p>`)}
    <p>This should stay</p>
  </div>
`);
```

The _udomdiff_ `before` element here becomes handy because it will work on this tree:
```html
<div>
  <!--pin-->
  <p>This should stay</p>
</div>
```
The list of items can grow and shrink at any time and nothing will ever be appended to that container.

```html
<div>
  <p>first</p>
  <p>second</p>
  <!--pin-->
  <p>This should stay</p>
</div>
```

However, I see that even enabling the `before` in the current refactoring, all libraries seems to work just fine, as they use `nextSibling` anyway, so they work in a clean way, which is why I am not using `before` anymore, but all libraries could receive it, either as undefined/null or as a proper node.
